File: java/src/Solution1143.java
public class Solution1143 {
	public int longestCommonSubsequence(String text1, String text2) {
		int m = text1.length(), n = text2.length();
		text1 = " " + text1;
		text2 = " " + text2;

		int[][] f = new int[m + 1][n + 1];

		for (int i = 0; i < m; i++) {
			f[i][0] = 1;
		}

		for (int j = 0; j < n; j++) {
			f[0][j] = 1;
		}

		for (int i = 1; i <= m; i++) {
			for (int j = 1; j <= n; j++) {
				if (text1.charAt(i) == text2.charAt(j)) {
					f[i][j] = f[i-1][j-1] + 1;
				} else {
					f[i][j] = Math.max(f[i-1][j], f[i][j-1]);
				}
			}
		}
		return f[m][n] - 1;
	}
}

File: golangR2/solution/solution1143.go
package solution

// dp[i][j]表示text1的前i个，text2的前j个的最长公共子序列
func LongestCommonSubsequence(text1 string, text2 string) int {
	m, n := len(text1), len(text2)

	dp := make([][]int, m+1)
	for i := 0; i <= m; i++ {
		dp[i] = make([]int, n+1)
	}

	for i := 0; i <= m; i++ {
		dp[i][0] = 0
	}
	for j := 0; j <= n; j++ {
		dp[0][j] = 0
	}

	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			if text1[i-1] == text2[j-1] {
				dp[i][j] = dp[i-1][j-1] + 1
			} else {
				if dp[i-1][j] > dp[i][j-1] {
					dp[i][j] = dp[i-1][j]
				} else {
					dp[i][j] = dp[i][j-1]
				}
			}
		}
	}

	return dp[m][n]
}
